### PR TITLE
[MPS] Migrate isposinf and isneginf to Metal kernels

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
# [MPS] Migrate isposinf and isneginf to Metal kernels

Part of the MPSGraph shape-keyed graph-cache elimination effort (#187455). Replaces the shared MPSGraph helper behind the `isposinf`/`isneginf` stubs with trivial Metal functors on the `exec_unary_kernel` path (bool output; integral inputs are never infinite), removing another shape-keyed graph-cache entry. The yaml rows were already shared, so only the stub kernels change.

Both ops opt into the existing ILP dense variant at the 256K crossover (bool output is non-floating, so ILP is opt-in there — same as `logical_not`).

## Perf

`benchmarks/mps/bench_isinf.py` (Timer.blocked_autorange, K=20 amortized behind one synchronize, median of 3 interleaved A/B reps vs stock nightly), M4 Pro — 20/20 cells, 0 regressions:

| cell | dtype | shape | speedup |
|---|---|---|---|
| isposinf | f32 | (4096,) | 2.45x |
| isneginf | f32 | (4096,) | 2.41x |
| isposinf | f32 | (512, 768) | 1.81x |
| isposinf | f32 | (2048, 2048) | 1.02x |
| isneginf | f32 | (16, 128, 1024) | 0.99x |
| (remaining 15 cells) | | | 0.99–2.42x |

## Test plan

```
python -m pytest test/test_mps.py -k "isposinf or isneginf or isinf"
python -m pytest test/test_ops.py -k "mps and (isposinf or isneginf)"
python benchmarks/mps/bench_isinf.py
```

Full `--mps` matrix run locally: only failure is the known `test_metal` desktop-build artifact. Standalone smoke: inf/-inf/NaN/0 edges across f32/f16/bf16, integral dtypes, a 2^19-element ILP-path tensor, transposed inputs, and `out=`.

Review log: https://gist.github.com/anagnorisis2peripeteia/5d22bf3dc629d5bdb68d621a4570b50c

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/TensorCompare.metal [hunk 1]
```diff
@@ -53,3 +53,42 @@ REGISTER_ALL_CLAMP_OPS(bool);
...
+  template <typename T, enable_if_t<is_scalar_integral_v<T>, bool> = true>
+  inline bool operator()(const T) {
+    return false;
+  }
+};
+
+struct isneginf_functor {
+  template <typename T, enable_if_t<is_scalar_floating_point_v<T>, bool> = true>
+  inline bool operator()(const T x) {
+    return float(x) == -INFINITY;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/TensorCompare.metal]: Kind of an inefficient cast for bf16, isn't it?

File: aten/src/ATen/native/mps/kernels/TensorCompare.metal [hunk 2]
```diff
@@ -53,3 +53,42 @@ REGISTER_ALL_CLAMP_OPS(bool);
...
+    return float(x) == INFINITY;
+  }
+  template <typename T, enable_if_t<is_scalar_integral_v<T>, bool> = true>
+  inline bool operator()(const T) {
+    return false;
+  }
+};
+
+struct isneginf_functor {
+  template <typename T, enable_if_t<is_scalar_floating_point_v<T>, bool> = true>
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/TensorCompare.metal]: Shouldn't this enable_of_t be based on hasQuietNaN std?

Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/kernels/TensorCompare.metal]: Good call — switched the dispatch to key on `::metal::numeric_limits<T>::has_infinity` (Metal's numeric_limits covers float/half/bfloat with `has_infinity == true` and integral/bool with `false`), which states the actual property the functor cares about rather than the float/int split.

File: aten/src/ATen/native/mps/kernels/TensorCompare.metal (hunk 1 above)
Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/kernels/TensorCompare.metal]: Agreed — the comparison now happens in the native type against `numeric_limits<T>::infinity()`, so half/bfloat no longer widen to float. Verified the half-max edge (±65504 stays non-inf) and re-ran the isposinf/isneginf suites.

**Comment by anagnorisis2peripeteia** [comment]
@Skylion007 is this flaky checks also?

**Comment by anagnorisis2peripeteia** [comment]
@Skylion007 sorry bumping this again

File: aten/src/ATen/native/mps/kernels/TensorCompare.metal [hunk 3]
```diff
@@ -53,3 +53,50 @@ REGISTER_ALL_CLAMP_OPS(bool);
...
+
+REGISTER_ISINF_OPS(float);
+REGISTER_ISINF_OPS(half);
+REGISTER_ISINF_OPS(bfloat);
+REGISTER_ISINF_OPS(long);
+REGISTER_ISINF_OPS(int);
+REGISTER_ISINF_OPS(short);
+REGISTER_ISINF_OPS(char);
+REGISTER_ISINF_OPS(uchar);
+REGISTER_ISINF_OPS(bool);
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/TensorCompare.metal]: Those are not needed because
https://github.com/pytorch/pytorch/blob/e289091a3a95a1987040333bcd45b039a69702df/aten/src/ATen/native/TensorCompare.cpp#L992-L998

**Review by malfet — APPROVED** [review]
LGTM, but can you check for the remaining PRs if you are making similar mistakes in your codebase (i.e. registering kernels that would never be dispatched to)

**Comment by anagnorisis2peripeteia** [comment]
@malfet checked the rest of my open MPS PRs for this. Short answer: no other instances.

The rule I applied is that a registration is dead only if the ATen-level impl never routes that
dtype to the stub. Here it was the `isIntegralType(..., includeBool=true)` short-circuit in
`isposinf_out`/`isneginf_out`, which fills the result and never calls `isposinf_stub`.

Across the open non-draft PRs:

* Float-only ops register float-only. layer_norm (#182109), the softmax family (#187456, #187457,
  #189146), cross_entropy (#187458), the loss leaves (#187553, #187556, #187557, #187559) and prelu
  (#189703) instantiate `float`/`half`/`bfloat` only.
* The broad lists are reachable rather than dead. `where` (#189627) registers `uint`/`ulong`/`ushort`,
  which `getMPSScalarType` maps to UInt16/32/64, and the complex rows in #189490 and #189420 map to
  `MPSDataTypeComplexFloat16`/`32`.
* The reduction PRs (#187787, #188156, #188333) use accumulate-type mappings rather than extra
  input dtypes.
* #187976 only touches `c10/metal/reduction_utils.h` and a test, so it registers nothing.

One near miss is worth recording, because it nearly went the other way. `round`'s integral rows look
dead from `round_out` alone, since `CREATE_UNARY_TORCH_IMPL_INTEGER_NO_OP_FUNC` does
`result.copy_(self)` for integral input. But `round_decimals_out` also calls `round_stub` in its
`decimals == 0` branch with no integral guard, and `TORCH_META_FUNC2(round, decimals)` builds a plain
unary op with no dtype restriction, so `torch.round(int_tensor, decimals=0)` does dispatch them and
they have to stay. Enumerating every caller of the stub, not just the identically named `_out`, is
what separates that case from the isposinf one.

This is static analysis of the dispatch paths rather than a device run, so treat it as such.

That should leave the stack clean on this front, and hopefully all ready for review when you have time.
